File: semestre-1/disciplina-2-modelagem-de-dados/unidade-4/aula-1/aula-1.py
import pandas as pd

def criar_tabela_original():
    """
    Cria e retorna um DataFrame do pandas representando a tabela original
    com dados não-normalizados do formulário do clube de futebol.
    Esta tabela contém redundâncias e dependências funcionais mal definidas.
    """
    dados = {
        'NumeroJogo': [101, 102, 103, 104, 105],
        'DataPartida': ['2023-10-25', '2023-11-02', '2023-11-10', '2023-11-18', '2023-11-25'],
        'LocalJogo': ['Estádio Municipal', 'Arena Olímpica', 'Estádio Municipal', 'Estádio do Morumbi', 'Arena Olímpica'],
        'CidadeLocal': ['Santa Catarina', 'Rio de Janeiro', 'Santa Catarina', 'São Paulo', 'Rio de Janeiro'],
        'NomeOponente': ['Time A', 'Time B', 'Time C', 'Time D', 'Time E'],
        'CodigoJogador': ['J001', 'J002', 'J001', 'J003', 'J002'],
        'NomeJogador': ['João Silva', 'Maria Santos', 'João Silva', 'Pedro Rocha', 'Maria Santos'],
        'PosicaoJogador': ['Atacante', 'Meio-campo', 'Atacante', 'Defensor', 'Meio-campo'],
        'GolsMarcados': [2, 1, 0, 3, 1]
    }
    return pd.DataFrame(dados)

def normalizar_para_primeira_forma_normal(df_original):
    """
    A 1ª Forma Normal (1FN) garante que cada célula contenha um único valor atômico.
    Neste exemplo, a tabela original já está em 1FN.
    """
    print("--- 1ª FORMA NORMAL (1FN) ---")
    print("A tabela original já atende à 1FN, pois não há grupos repetitivos ou listas em uma única célula.")
    return df_original.copy()

def normalizar_para_segunda_forma_normal(df_1fn):
    """
    A 2ª Forma Normal (2FN) remove dependências parciais.
    Nossa chave primária é a combinação de (NumeroJogo, CodigoJogador).
    'NomeJogador' e 'PosicaoJogador' dependem apenas de 'CodigoJogador', não da chave inteira.
    'LocalJogo' e 'CidadeLocal' dependem apenas de 'NumeroJogo', não da chave inteira.
    Vamos separar essas informações em novas tabelas.
    """
    print("\n--- 2ª FORMA NORMAL (2FN) ---")
    
    # Tabela 1: JOGOS (Dependência de NumeroJogo)
    tabela_jogos = df_1fn[['NumeroJogo', 'DataPartida', 'LocalJogo', 'CidadeLocal', 'NomeOponente']].drop_duplicates().reset_index(drop=True)
    print("\n[Tabela JOGOS] - Removendo dependência parcial de 'LocalJogo' e 'CidadeLocal'.")
    print(tabela_jogos)

    # Tabela 2: JOGADORES (Dependência de CodigoJogador)
    tabela_jogadores = df_1fn[['CodigoJogador', 'NomeJogador', 'PosicaoJogador']].drop_duplicates().reset_index(drop=True)
    print("\n[Tabela JOGADORES] - Removendo dependência parcial de 'NomeJogador' e 'PosicaoJogador'.")
    print(tabela_jogadores)
    
    # Tabela 3: ATUACOES (Tabela de ligação, com a chave primária composta)
    tabela_atuacoes = df_1fn[['NumeroJogo', 'CodigoJogador', 'GolsMarcados']].copy()
    print("\n[Tabela ATUACOES] - Tabela de ligação que mantém o relacionamento entre JOGOS e JOGADORES.")
    print(tabela_atuacoes)
    
    return tabela_jogos, tabela_jogadores, tabela_atuacoes

def normalizar_para_terceira_forma_normal(tabela_jogos):
    """
    A 3ª Forma Normal (3FN) remove dependências transitivas.
    Em nossa tabela 'JOGOS', o 'CidadeLocal' depende de 'LocalJogo', que não é a chave primária.
    Vamos separar essa informação em uma nova tabela de LOCALIZACAO.
    """
    print("\n--- 3ª FORMA NORMAL (3FN) ---")
    
    # Tabela 1: LOCALIZACOES (Dependência transitiva de CidadeLocal)
    tabela_localizacoes = tabela_jogos[['LocalJogo', 'CidadeLocal']].drop_duplicates().reset_index(drop=True)
    print("\n[Tabela LOCALIZACOES] - Removendo a dependência transitiva de 'CidadeLocal'.")
    print(tabela_localizacoes)

    # Tabela 2: JOGOS_NORMALIZADA (Tabela de jogos agora referenciando a nova tabela)
    tabela_jogos_normalizada = tabela_jogos.drop(columns=['CidadeLocal']).copy()
    print("\n[Tabela JOGOS_NORMALIZADA] - Tabela JOGOS após a normalização.")
    print(tabela_jogos_normalizada)

    return tabela_localizacoes, tabela_jogos_normalizada


# --- Execução do processo de Normalização ---
if __name__ == "__main__":
    print("--------------------------------------------------")
    print("      EXEMPLO DE NORMALIZAÇÃO DE DADOS EM PYTHON     ")
    print("--------------------------------------------------")

    # Passo 1: Criação da tabela não-normalizada
    df_formulario = criar_tabela_original()
    print("\n[Tabela Original - Formulário do Clube]")
    print(df_formulario)
    
    # Passo 2: Normalização para 1FN
    df_1fn = normalizar_para_primeira_forma_normal(df_formulario)
    
    # Passo 3: Normalização para 2FN
    jogos, jogadores, atuacoes = normalizar_para_segunda_forma_normal(df_1fn)
    
    # Passo 4: Normalização para 3FN
    localizacoes, jogos_normalizada = normalizar_para_terceira_forma_normal(jogos)

    print("\n--------------------------------------------------")
    print("           PROCESSO DE NORMALIZAÇÃO FINALIZADO       ")
    print("--------------------------------------------------")
    print("Observe como o processo de normalização dividiu uma única tabela com redundâncias em tabelas menores, mais organizadas e eficientes, eliminando a repetição de dados.")

File: semestre-1/disciplina-2-modelagem-de-dados/unidade-4/aula-1/aula-1.java
// aula-1.java

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

/**
 * Classe para simular a tabela original, não normalizada,
 * com dados de um formulário de registro de jogos de futebol.
 */
class RegistroJogo {
    int numeroJogo;
    String dataPartida;
    String localJogo;
    String cidadeLocal;
    String nomeOponente;
    String codigoJogador;
    String nomeJogador;
    String posicaoJogador;
    int golsMarcados;

    public RegistroJogo(int numeroJogo, String dataPartida, String localJogo, String cidadeLocal, String nomeOponente, String codigoJogador, String nomeJogador, String posicaoJogador, int golsMarcados) {
        this.numeroJogo = numeroJogo;
        this.dataPartida = dataPartida;
        this.localJogo = localJogo;
        this.cidadeLocal = cidadeLocal;
        this.nomeOponente = nomeOponente;
        this.codigoJogador = codigoJogador;
        this.nomeJogador = nomeJogador;
        this.posicaoJogador = posicaoJogador;
        this.golsMarcados = golsMarcados;
    }

    @Override
    public String toString() {
        return "RegistroJogo{" +
                "numeroJogo=" + numeroJogo +
                ", dataPartida='" + dataPartida + '\'' +
                ", localJogo='" + localJogo + '\'' +
                ", cidadeLocal='" + cidadeLocal + '\'' +
                ", nomeOponente='" + nomeOponente + '\'' +
                ", codigoJogador='" + codigoJogador + '\'' +
                ", nomeJogador='" + nomeJogador + '\'' +
                ", posicaoJogador='" + posicaoJogador + '\'' +
                ", golsMarcados=" + golsMarcados +
                '}';
    }
}

/**
 * Representa a tabela normalizada 'Jogos'.
 */
class Jogo {
    int numeroJogo;
    String dataPartida;
    String localJogo;
    String nomeOponente;

    public Jogo(int numeroJogo, String dataPartida, String localJogo, String nomeOponente) {
        this.numeroJogo = numeroJogo;
        this.dataPartida = dataPartida;
        this.localJogo = localJogo;
        this.nomeOponente = nomeOponente;
    }

    @Override
    public String toString() {
        return "Jogo{" +
                "numeroJogo=" + numeroJogo +
                ", dataPartida='" + dataPartida + '\'' +
                ", localJogo='" + localJogo + '\'' +
                ", nomeOponente='" + nomeOponente + '\'' +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Jogo jogo = (Jogo) o;
        return numeroJogo == jogo.numeroJogo;
    }

    @Override
    public int hashCode() {
        return Objects.hash(numeroJogo);
    }
}

/**
 * Representa a tabela normalizada 'Jogadores'.
 */
class Jogador {
    String codigoJogador;
    String nomeJogador;
    String posicaoJogador;

    public Jogador(String codigoJogador, String nomeJogador, String posicaoJogador) {
        this.codigoJogador = codigoJogador;
        this.nomeJogador = nomeJogador;
        this.posicaoJogador = posicaoJogador;
    }

    @Override
    public String toString() {
        return "Jogador{" +
                "codigoJogador='" + codigoJogador + '\'' +
                ", nomeJogador='" + nomeJogador + '\'' +
                ", posicaoJogador='" + posicaoJogador + '\'' +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Jogador jogador = (Jogador) o;
        return codigoJogador.equals(jogador.codigoJogador);
    }

    @Override
    public int hashCode() {
        return Objects.hash(codigoJogador);
    }
}

/**
 * Representa a tabela de ligação 'Atuacoes'.
 */
class Atuacao {
    int numeroJogo;
    String codigoJogador;
    int golsMarcados;

    public Atuacao(int numeroJogo, String codigoJogador, int golsMarcados) {
        this.numeroJogo = numeroJogo;
        this.codigoJogador = codigoJogador;
        this.golsMarcados = golsMarcados;
    }

    @Override
    public String toString() {
        return "Atuacao{" +
                "numeroJogo=" + numeroJogo +
                ", codigoJogador='" + codigoJogador + '\'' +
                ", golsMarcados=" + golsMarcados +
                '}';
    }
}

/**
 * Representa a tabela normalizada 'Localizacoes'.
 */
class Localizacao {
    String localJogo;
    String cidadeLocal;

    public Localizacao(String localJogo, String cidadeLocal) {
        this.localJogo = localJogo;
        this.cidadeLocal = cidadeLocal;
    }

    @Override
    public String toString() {
        return "Localizacao{" +
                "localJogo='" + localJogo + '\'' +
                ", cidadeLocal='" + cidadeLocal + '\'' +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Localizacao that = (Localizacao) o;
        return localJogo.equals(that.localJogo);
    }

    @Override
    public int hashCode() {
        return Objects.hash(localJogo);
    }
}

public class aula_1 {

    public static void main(String[] args) {
        System.out.println("--------------------------------------------------");
        System.out.println("    EXEMPLO DE NORMALIZAÇÃO DE DADOS EM JAVA    ");
        System.out.println("--------------------------------------------------");

        // Passo 1: Criação da tabela original não normalizada
        List<RegistroJogo> tabelaOriginal = criarTabelaOriginal();
        System.out.println("\n[Tabela Original - Formulário do Clube]");
        tabelaOriginal.forEach(System.out::println);

        // Passo 2: Normalização para 2FN (já que a 1FN é atendida)
        System.out.println("\n--- Normalização para 2ª Forma Normal (2FN) ---");
        System.out.println("Separando as dependências parciais de 'CodigoJogador' e 'NumeroJogo'.");

        // Cria a tabela de Jogos e remove duplicatas (com base no numeroJogo)
        List<Jogo> tabelaJogos = tabelaOriginal.stream()
                .map(r -> new Jogo(r.numeroJogo, r.dataPartida, r.localJogo, r.nomeOponente))
                .distinct()
                .collect(Collectors.toList());

        // Cria a tabela de Jogadores e remove duplicatas (com base no codigoJogador)
        List<Jogador> tabelaJogadores = tabelaOriginal.stream()
                .map(r -> new Jogador(r.codigoJogador, r.nomeJogador, r.posicaoJogador))
                .distinct()
                .collect(Collectors.toList());

        // Cria a tabela de Atuações (tabela de ligação)
        List<Atuacao> tabelaAtuacoes = tabelaOriginal.stream()
                .map(r -> new Atuacao(r.numeroJogo, r.codigoJogador, r.golsMarcados))
                .collect(Collectors.toList());

        System.out.println("\n[Tabela JOGOS]");
        tabelaJogos.forEach(System.out::println);
        System.out.println("\n[Tabela JOGADORES]");
        tabelaJogadores.forEach(System.out::println);
        System.out.println("\n[Tabela ATUACOES]");
        tabelaAtuacoes.forEach(System.out::println);

        // Passo 3: Normalização para 3FN
        System.out.println("\n--- Normalização para 3ª Forma Normal (3FN) ---");
        System.out.println("Separando a dependência transitiva de 'CidadeLocal' de 'LocalJogo'.");

        // Cria a tabela de Localizacoes e remove duplicatas (com base no localJogo)
        List<Localizacao> tabelaLocalizacoes = tabelaOriginal.stream()
                .map(r -> new Localizacao(r.localJogo, r.cidadeLocal))
                .distinct()
                .collect(Collectors.toList());
        
        // A tabela de Jogos original, agora com a dependência transitiva removida, já foi criada.

        System.out.println("\n[Tabela LOCALIZACOES]");
        tabelaLocalizacoes.forEach(System.out::println);
        System.out.println("\n[Tabela JOGOS (após a 3FN)]");
        tabelaJogos.forEach(System.out::println);

        System.out.println("\n--------------------------------------------------");
        System.out.println("          PROCESSO DE NORMALIZAÇÃO FINALIZADO       ");
        System.out.println("--------------------------------------------------");
    }

    private static List<RegistroJogo> criarTabelaOriginal() {
        List<RegistroJogo> registros = new ArrayList<>();
        registros.add(new RegistroJogo(101, "2023-10-25", "Estádio Municipal", "Santa Catarina", "Time A", "J001", "João Silva", "Atacante", 2));
        registros.add(new RegistroJogo(102, "2023-11-02", "Arena Olímpica", "Rio de Janeiro", "Time B", "J002", "Maria Santos", "Meio-campo", 1));
        registros.add(new RegistroJogo(103, "2023-11-10", "Estádio Municipal", "Santa Catarina", "Time C", "J001", "João Silva", "Atacante", 0));
        registros.add(new RegistroJogo(104, "2023-11-18", "Estádio do Morumbi", "São Paulo", "Time D", "J003", "Pedro Rocha", "Defensor", 3));
        registros.add(new RegistroJogo(105, "2023-11-25", "Arena Olímpica", "Rio de Janeiro", "Time E", "J002", "Maria Santos", "Meio-campo", 1));
        return registros;
    }
}